### PR TITLE
[hotfix] Suport extra_kwargs in ShardConfig

### PR DESCRIPTION
**Pull request by KKZ20**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Refactor the shard config: support using extra_kwargs in ShardConfig

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [x] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/inference/hybridengine/polices/llama.py [hunk 1]
```diff
@@ -45,8 +45,10 @@ def __init__(self) -> None:
 
     def module_policy(self):
         policy = super().module_policy()
-
-        if self.shard_config.inference_gptq:
+        if (
+            "inference_gptq" in self.shard_config.extra_kwargs.keys()
```
Comment (ver217) [inline colossalai/inference/hybridengine/polices/llama.py:49]: You can use `self.shard_config.extra_kwargs.get("inference_gptq", False)`

File: colossalai/inference/tensor_parallel/engine.py [hunk 2]
```diff
@@ -221,13 +221,17 @@ def _shard_model_by(self, shardformer: ShardFormer, model: nn.Module) -> None:
         ), "Discrepancy between the tp size of TPInferEngine and the tp size of shard config"
         model_name = model.__class__.__name__
         assert model_name in self.supported_models, f"Unsupported model cls {model_name} for TP inference."
-
-        model = model.model if self.shard_config.inference_gptq else model
+        if (
+            "inference_gptq" in self.shard_config.extra_kwargs.keys()
```
Comment (ver217) [inline colossalai/inference/tensor_parallel/engine.py:225]: Same

File: colossalai/inference/tensor_parallel/engine.py [hunk 3]
```diff
@@ -221,13 +221,17 @@ def _shard_model_by(self, shardformer: ShardFormer, model: nn.Module) -> None:
...
+            and self.shard_config.extra_kwargs["inference_gptq"]
+        ):
+            model = model.model
         policy = get_autopolicy(model, shard_config=self.shard_config)
-
         self.model, _ = shardformer.optimize(model, policy)
-
-        if self.shard_config.inference_gptq:
+        if (
+            "inference_gptq" in self.shard_config.extra_kwargs.keys()
```
Comment (ver217) [inline colossalai/inference/tensor_parallel/engine.py:232]: Same

File: colossalai/inference/tensor_parallel/policies/bloom.py [hunk 4]
```diff
@@ -38,35 +37,42 @@ def module_policy(self):
         from transformers.models.bloom.modeling_bloom import BloomAttention, BloomBlock, BloomForCausalLM, BloomModel
 
         policy = super().module_policy()
-        if self.shard_config.inference_gptq:
+
+        if (
+            "inference_gptq" in self.shard_config.extra_kwargs.keys()
```
Comment (ver217) [inline colossalai/inference/tensor_parallel/policies/bloom.py:42]: same

File: colossalai/inference/tensor_parallel/policies/llama.py [hunk 5]
```diff
@@ -36,7 +38,10 @@ def __init__(self) -> None:
     def module_policy(self):
         policy = super().module_policy()
 
-        if self.shard_config.inference_gptq:
+        if (
+            "inference_gptq" in self.shard_config.extra_kwargs.keys()
```
Comment (ver217) [inline colossalai/inference/tensor_parallel/policies/llama.py:42]: same

File: colossalai/shardformer/shard/shard_config.py [hunk 6]
```diff
@@ -46,6 +44,10 @@ def tensor_parallel_size(self):
         return self._tensor_parallel_size
 
     def __post_init__(self):
+        if self.extra_kwargs:
+            for key, value in self.extra_kwargs.items():
+                setattr(self, key, value)
```
Comment (ver217) [inline colossalai/shardformer/shard/shard_config.py:49]: Is it essential to copy the attr back? If so you have to prevent conflicts here

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/shard/shard_config.py (hunk 6 above)
Comment (KKZ20) [reply colossalai/shardformer/shard/shard_config.py]: Removed

File: colossalai/inference/tensor_parallel/policies/llama.py (hunk 5 above)
Comment (KKZ20) [reply colossalai/inference/tensor_parallel/policies/llama.py]: Fixed.

File: colossalai/inference/tensor_parallel/policies/bloom.py (hunk 4 above)
Comment (KKZ20) [reply colossalai/inference/tensor_parallel/policies/bloom.py]: Fixed.

File: colossalai/inference/tensor_parallel/engine.py (hunk 3 above)
Comment (KKZ20) [reply colossalai/inference/tensor_parallel/engine.py]: Fixed.

File: colossalai/inference/tensor_parallel/engine.py (hunk 2 above)
Comment (KKZ20) [reply colossalai/inference/tensor_parallel/engine.py]: Fixed.

File: colossalai/inference/hybridengine/polices/llama.py (hunk 1 above)
Comment (KKZ20) [reply colossalai/inference/hybridengine/polices/llama.py]: Fixed.

**Review by FrankLeeeee — APPROVED** [review]

**Review by ver217 — APPROVED** [review]

**Pull request merged**